### PR TITLE
Improve backtrace abstractions inside runtime

### PR DESCRIPTION
**Pull request by NickBarnes**
Note: this is not a PR on ocaml/ocaml yet, as I want to make several statmemprof-related PRs there together, so that the motivation for this work is clearly linked to statmemprof. So I've made a PR on my fork repo to gather feedback from some domain experts.

In the trunk runtime, the backtrace API allows backtraces to be obtained either as (a) a single per-domain buffer, reserved for the "backtrace at the last exception raise", or (b) an object on the Caml heap, reflecting the current backtrace. This is insufficient for statmemprof, which needs the current backtrace at arbitrary allocation points, when Caml heap allocation might not be possible (or might change the backtrace, by garbage collection). So this branch changes the `backtrace.h` abstraction by adding `caml_get_callstack()`.

I took the opportunity to also improve the code which actually iterates over the stack gathering backtrace entries—in each runtime (native-code and bytecode) there's now a single piece of code which does this, and it only iterates over the stack once. This key function (`get_callstack()` in both `backtrace_nat.c` and `backtrace_byt.c`) has a rather ugly ("candy machine") interface, because it turns out that the semantics of this are pretty twisty, depending for instance on whether the callstack should trace up into parent stack fibers or just to the current fiber, which depends (in the native-code runtime) on exactly why we want a backtrace. I'm not sure the current semantics are all that great but I've tried to preserve them.

I was tempted to improve the abstraction further, by making a new type (`callstack_t`) of C-heap-allocated backtrace objects which have slots for some metadata such as frame count (number of live entries) and size (for re-sizing), and support a small number of operations. That would allow this code to be further improved, removing some duplicated code in each runtime and also moving some source code from `backtrace_nat.c` and `backtrace_byt.c` into `backtrace.c`. However, I've deferred that for now.

### CODE REVIEW

File: runtime/backtrace_nat.c [hunk 1]
```diff
@@ -81,16 +84,86 @@ void caml_free_backtrace_buffer(backtrace_slot *backtrace_buffer) {
...
+static debuginfo debuginfo_extract(frame_descr *d, ssize_t alloc_idx);
+
+/* initial size of a callstack buffer, in entries */
+
+#define MIN_CALLSTACK_SIZE 16
+
+/* Stores up to [max_frames] frame descriptors of a stack in a buffer
+   [*trace_p] of length [*allocated_size_p], allocated in the C
+   heap. Resize the buffer as required. Start at [sp]/[pc] on [stack],
+   and at offfset [frames] in the buffer. Stop when we reach
```
Comment (fabbing) [inline runtime/backtrace_nat.c:103]:
```suggestion
   and at offset [frames] in the buffer. Stop when we reach
```

File: runtime/backtrace_nat.c [hunk 2]
```diff
@@ -102,127 +175,118 @@ void caml_stash_backtrace(value exn, uintnat pc, char * sp, char* trapsp)
...
+  backtrace_slot *backtrace_buffer = domain_state->backtrace_buffer;
+  domain_state->backtrace_pos = get_callstack(domain_state->current_stack,
+                                              sp, pc, trapsp,
+                                              false, -1,
+                                              BACKTRACE_BUFFER_SIZE,
+                                              domain_state->backtrace_pos,
+                                              &backtrace_buffer,
+                                              &allocated_size);
+  CAMLassert(domain_state->backtrace_pos <= BACKTRACE_BUFFER_SIZE);
+  CAMLassert(allocated_size == BACKTRACE_BUFFER_SIZE);
```
Comment (fabbing) [inline runtime/backtrace_nat.c:190]: I'm not sure to understand what enforce this assertion.
The allocation looks "dynamic" in `get_callstack` with the call to `caml_stat_resize_noexc` and independent of `BACKTRACE_BUFFER_SIZE` / `max_frames`. What did I miss?

**Review by fabbing — COMMENTED** [review]
Looks good to me!
That's indeed a lot of `parameters`/`options` for `get_callstack` but the function itself is still very readable.

File: runtime/backtrace_nat.c (hunk 2 above)
Comment (NickBarnes) [reply runtime/backtrace_nat.c:190]: The point is that we give `get_callstack` a buffer of size `BACKTRACE_BUFFER_SIZE` and tell it not to return more than that number of frames, so it shouldn't ever resize. I'll add a comment.

In an ideal world, the `domain_state` wouldn't have a huge buffer of fixed size, but would create and grow its buffer as required, but that's a bigger change than I wanted to make in this PR.

**Pull request closed without merging**